synapse/assets/storm/migrations/model-0.2.27.storm

```diff
@@ -0,0 +1,356 @@
+                ($verb, $src) = $edge
+                $edges.append($edge)
+
+                [ <($verb)- { iden $src } ]
+
+                // Flush to the queue if the list grows too large
+                if ($edges.size() > 1000) { // pragma: no cover start
+                    $item = ({
+                        "iden": $iden,
+                        "view": $view.iden,
+                        "direction": "n2",
+                        "edges": $edges,
+                    })
+
+                    $edgeq.put($item)
+                    $edges = ([])
+                } // pragma: no cover stop
+            }}
+
+            if $edges {
+                $item = ({
+                    "iden": $iden,
+                    "view": $view.iden,
+                    "direction": "n2",
+                    "edges": $edges,
+                })
+
+                $edgeq.put($item)
+                $edges = ([])
+            }
+        }
+    }
+
+    $item = ({
+        "iden": $iden,
+        "form": $form,
+        "repr": $repr,
+        "props": $n.props,
+        "view": $viewiden,
+        "layer": $layriden,
+        "tags": $n.tags(),
+        "data": $n.data.list(),
+        "sources": $sources,
+    })
+
+    $nodeq.put($item)
+
+    return()
+}
+
+function migrateCpeNode(oldcpe, newcpe) {
+
+    // migrate the edges, tags, data, and universal properties
+    $lib.model.migration.copyEdges($oldcpe, $newcpe)
+    $lib.model.migration.copyTags($oldcpe, $newcpe)
+    $lib.model.migration.copyData($oldcpe, $newcpe)
+
+    $refinfos = $getRefInfo($oldcpe.form())
+
+    // Iterate through the views and fix up all the references
+    for $view in $lib.view.list(deporder=$lib.true) {
+        if (not $layers.has($view.layers.0.iden)) { continue }
+
+        view.exec $view.iden {
+
+            // Fix references that point to old node to now point to new node
+            for $refinfo in $refinfos {
+                ($form, $prop, $proptype, $isarray) = $refinfo
+                if ($proptype = "ndef") {
+                    $oldvalu = $oldcpe.ndef()
+                    $newvalu = $newcpe.ndef()
+                } else {
+                    $oldvalu = $oldcpe.repr()
+                    $newvalu = $newcpe.repr()
+                }
+
+                for $n in $lib.model.migration.liftByPropValuNoNorm($form, $prop, $oldvalu) {
+                    yield $n
+
+                    if $isarray {
+
+                        // We can't just [ :$prop-=$oldvalu :$prop+=$newvalu ] because the norm() function gets called
+                        // on the array type deep down in the AST. So, instead, we have to operate on the whole array.
+
+                        $list = $lib.copy(:$prop)
+                        while $list.has($oldvalu) {
+                            $list.rem($oldvalu)
+                        }
+
+                        if (not $list.has($newvalu)) {
+                            $list.append($newvalu)
```

Comment:
Might be getting too far into edge cases here but this will change the order of unsorted array props and could remove multiple but only re-add one instance in non-uniq array props

Reply:
It's a list so the order shouldn't change. The while block above will definitely remove all the invalid old values if they're in the list more than once. This block will just append the new value once. I'm not sure I understand your concern.

Reply:
ping @Cisphyx ^^^

Reply:
Made up example of a non-uniq, unsorted array prop (doesn't seem useful which is why i noted this was likely an edge case we can ignore):
`:cpelist=(CPE-1, CPE-2, CPE-2, CPE-3)`
Say CPE-2 was invalid and now is CPE-4, you would end up with:
`:cpelist=(CPE-1, CPE-3, CPE-4)`
When if for some reason the order and duplicate values mattered, it should be:
`:cpelist=(CPE-1, CPE-4, CPE-4, CPE-3)`